### ISSUE TITLE
Block requests made by inactive tabs

### ISSUE THREAD

**Issue by teodor6140** [uBlockOrigin/uBlock-issues#646]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [x] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description
Add a feature to block requests made by background tabs, allow whitelisting for some urls. Not sure if this belongs here, let me know if there's an extension that does this. There's one called "Auto Tab Discard" which unloads inactive tabs completely, but I would like to keep the tab loaded and block requests made by it to prevent tracking.

**Comment by gorhill** [comment]
> I would like to keep the tab loaded and block requests made by it to prevent tracking.

There are lists selected by default in uBO to prevent tracking. Can you elaborate what do you mean exactly?

**Comment by otzade** [comment]
Stricter blocking for background tabs, perhaps block all requests. And by tracking I mean background tabs which continuously make requests, this can be used for tracking amongst other things.

**Comment by gorhill** [comment]
> continuously make requests, this can be used for tracking amongst other things.

You are just repeating what you said. I want technical details of how tracking in a background tab is different than the same as tracking in a foreground tab -- keeping in mind uBO has many lists selected by default to minimize tracking.

**Comment by otzade** [comment]
I'm talking about requests which aren't primarily for tracking but can be used to track when you're online. Like a social media website polling for updates. The tracking filters don't block these because it would break the functionality provided by them, but there should be an option to disable such things in the background where they aren't as important.

**Comment by uBlock-user** [comment]
uBO doesn't block anything by default. You need either rules or filterlists which tell uBO what to block.

**Comment by uBlock-user** [comment]
What you're suggesting would be something more akin to uMatrix like extension.

**Comment by gorhill** [comment]
To be clear, I am not against the idea, I need actual real world cases demonstrating its benefits. I can't just add features because they sound nice, they have to be well rationalized and they have to have a chance of being used by more than a handful of users. Once this hurdle is cleared, I will have to evaluate the technical side as to whether the implementation is trivial and whether it has chance to cause regressions in existing features.

**Comment by gorhill** [comment]
Still waiting on:

> provide actual real world examples making the case of the benefits

**Comment by gorhill** [comment]
Closing due to lack of feedback, i.e. providing real world example where this is beneficial. Additionally, browser vendors are already implementing strategies to suspend unused tabs.

**Comment by uBlock-user** [comment]
Duplicate of https://github.com/gorhill/uBlock/issues/1118

**Comment by gorhill** [comment]
These sounds like real cases which would benefit from such feature:

- [Firefox (Android) used several GB of mobile data in background overnight](https://bugzilla.mozilla.org/show_bug.cgi?id=1273552)
- [Firefox uses huge amounts of mobile data](https://bugzilla.mozilla.org/show_bug.cgi?id=1565475)

**Comment by gwarser** [comment]
Just a note: there is difference between background tabs and Firefox app being in background on mobile.

**Comment by JKAbrams** [comment]
> Still waiting on:
> 
> > provide actual real world examples making the case of the benefits

On Firefox on Linux (I don't know if the problem also affects other browsers/OS but it is likely) I'm seeing a increasingly severe problem of tabs using lots and lots of CPU. The general state of JavaScript execution on the web seems have gone down quality wise.
I have not looked into why or studied what exactly what the browser is doing when it starts using CPU. For work I always have a large amount of tabs and windows open, it is not a problem as I have enough memory to handle it easily but CPU is a real problem as it starts eating it, after a (about a day or so) I find some random tab using 100% CPU.

As a user I have been wrecking my head trying to figure out what good all these CPU cycles does for me and on seemingly just static sites or dynamic sites that are just waiting, but except for the occasional site that I have allowed to do background announcements (like new messages on Discourse) I cannot answer this question and thus I like my having my battery power rather then having these scripts running. The solution for me seems obvious, I should try to find a way to deny these background scripts the privilege of using cycles of my machine. 

As I run uBlock in rather strict mode I have been wondering if this could make some scripts go haywire as the tracking scripts they expect to interact with are not available making them eat more resources than normal, this is just a thought which I don't have any proof to substantiate (my JavaScript debugging skills for others live websites are not good enough to go to the bottom with this) so I'll just leave it here as a thought if someone has more knowledge on the topic.

If indeed these scripts running in the background are not providing anything useful to their users, they are then serving somebody else's purpose and not the users.

As uBlock has added the feature to block JavaScript this kind of feature could be a really cool addition in this direction saving users massive amounts of wasted CPU cycles and helping users take back control of their computing resources, which I find as the essence of what uBlock is about.

As for the technical aspects of implementing this feature I do not know the plugin API well enough to know if JavaScript can be disabled for inactive tabs. By using the debugger there is a pause button that used to just freeze the website, but they added a grey overlay and it is no longer possible to scroll such a page, even tho it is already rendered in memory so at least the browser (at least Firefox) kind of has the ability to do this feature already.

Thanks for considering this issue.

**Comment by gwarser** [comment]
>JavaScript can be disabled for inactive tabs.

Cannot be done. You can use other extensions to "suspend" inactive or background tabs.

**Comment by gorhill** [comment]
Declining, I think it's best handled as a browser feature.
